pkg/policy/l4.go

```diff
@@ -1084,13 +1111,287 @@ func addL4Filter(policyCtx PolicyContext,
 		}
 	}
 
-	resMap[key] = existingFilter
+	resMap.Upsert(p.Port, uint16(p.EndPort), string(proto), existingFilter)
+	return nil
+}
+
+// L4PolicyMap is a list of L4 filters indexable by port/endport/protocol
+type L4PolicyMap interface {
+	Upsert(port string, endPort uint16, protocol string, l4 *L4Filter)
+	Delete(port string, endPort uint16, protocol string)
+	ExactLookup(port string, endPort uint16, protocol string) *L4Filter
+	LongestPrefixMatch(port string, protocol string) *L4Filter
+	Detach(selectorCache *SelectorCache)
+	IngressCoversContext(ctx *SearchContext) api.Decision
+	EgressCoversContext(ctx *SearchContext) api.Decision
+	ForEach(func(l4 *L4Filter) bool)
+	Equals(t *testing.T, bMap L4PolicyMap) bool
+	Diff(t *testing.T, expectedMap L4PolicyMap) string
```

Comment:
nit: We should probably look at which of these actually need to be public, and hide them by default. Upsert directly here rather than via `addL4Filter()` would bypass some rule merging logic, Delete could delete filters that are actively used by another rule, etc.

Reply:
[Jarno's PR](https://github.com/cilium/cilium/pull/33302) addresses this.

Reply:
I don't understand how that PR is related. The linked PR seems to be providing stronger guarantees at a subsequent policy stage related to EndpointPolicy and map state.

`addL4Filter()` does a merge L4filter + call into the publicly exported `Upsert` from this API. If ever someone in future decided to attempt to directly `Upsert` using this API in this interface, then they would be avoiding the filter merge logic. This is not a huge deal because if someone attempted to propose such a patch, we would also review that code and hopefully catch the discrepancy. I just point it out as a bit of a code smell that we shouldn't be exporting functions that are not intended for use outside the package.